src/app/course/_forms/course-event.form.ts

```diff
@@ -97,4 +117,127 @@ export class CourseEventForm {
+                    TuiDay.currentLocal(),
+                    TuiDay.currentLocal().append({day: 7})
+                ),
+                [Validators.required]
+            ),
+            startTimePicker: new FormControl(
+                TuiTime.currentLocal(),
+                [Validators.required]
+            ),
+            endTimePicker: new FormControl(
+                TuiTime.currentLocal(),
+                [Validators.required]
+            ),
+            challengeQuestionSets: new FormArray(
+                [CourseEventForm.createChallengeQuestionSetForm()],
+                [Validators.required]
+            )
+        },  {validator: CourseEventForm.dateValidator} as AbstractControlOptions)
+    }
+
+    static createChallengeFormWithData(challenge: CourseEvent): FormGroup {
+        const builder = new FormBuilder()
+        return builder.group({
+            name: new FormControl(challenge.name, [Validators.required]),
+            challengeType: new FormControl(
+                challenge.challenge_type,
+                [Validators.required]
+            ),
+            challengeTypeValue:  new FormControl(
+                challenge.challenge_type_value,
+                [Validators.required]
+            ),
+            maxTeamSize: new FormControl(
+                challenge.max_team_size,
+                [Validators.required]
+            ),
+            startEndDatePicker: new FormControl(
+                new TuiDayRange(
+                    TuiDay.fromLocalNativeDate(new Date(challenge.start_date)),
+                    TuiDay.fromLocalNativeDate(new Date(challenge.end_date))
+                ),
+                [Validators.required]
+            ),
+            startTimePicker: new FormControl(
+                TuiTime.fromLocalNativeDate(new Date(challenge.start_date)),
+                [Validators.required]
+            ),
+            endTimePicker: new FormControl(
+                TuiTime.fromLocalNativeDate(new Date(challenge.end_date)),
+                [Validators.required]
+            ),
+        }, {validator: CourseEventForm.dateValidator} as AbstractControlOptions)
+    }
+
+    static createChallengeQuestionSetForm(): FormGroup {
+        const builder = new FormBuilder()
+        return builder.group({
+            category: new FormControl('', [Validators.required]),
+            difficulty: new FormControl('', [Validators.required]),
+            number_of_questions: new FormControl('', [Validators.required])
+        })
+    }
+
+    //Need a new relationship with the Event model to relate the questions
+    // static createChallengeQuestionSetFromChallengeForm(challenge:CourseEvent): FormGroup {
+    //     const builder = new FormBuilder()
+    //     return builder.group({
+    //         category: new FormControl('', [Validators.required]),
+    //         difficulty: new FormControl('', [Validators.required]),
+    //         number_of_questions: new FormControl('', [Validators.required])
```

Comment:
I'm not sure how I can get those fields from an event object. It needs backend modification, but not sure where to start for the backend.